consensus/state_test.go

```diff
@@ -291,6 +294,51 @@ func TestStateFullRoundNil(t *testing.T) {
 	validatePrevoteAndPrecommit(t, cs, round, 0, vss[0], nil, nil)
 }
 
+// Tests that consensus state can be saved and reloaded from DB after committing a block.
+func TestStateFullRound1Reload(t *testing.T) {
+	blockDB := dbm.NewMemDB()
+	state, privVals := randGenesisState(1, false, 10)
+	vss := make([]*validatorStub, 1)
+	cs := newConsensusStateWithConfigAndBlockStore(config, state, privVals[0], counter.NewCounterApplication(true), blockDB)
+	vss[0] = NewValidatorStub(privVals[0], 0)
+	incrementHeight(vss[1:]...)
+	height, round := cs.Height, cs.Round
+
+	// NOTE: buffer capacity of 0 ensures we can validate prevote and last commit
+	// before consensus can move to the next height (and cause a race condition)
+	cs.eventBus.Stop()
+	eventBus := types.NewEventBusWithBufferCapacity(0)
+	eventBus.SetLogger(log.TestingLogger().With("module", "events"))
+	cs.SetEventBus(eventBus)
+	eventBus.Start()
+
+	voteCh := subscribe(cs.eventBus, types.EventQueryVote)
+	propCh := subscribe(cs.eventBus, types.EventQueryCompleteProposal)
+	newRoundCh := subscribe(cs.eventBus, types.EventQueryNewRound)
+
+	startTestRound(cs, height, round)
+
+	<-newRoundCh
+
+	// grab proposal
+	re := <-propCh
+	propBlockHash := re.(types.EventDataRoundState).RoundState.(*cstypes.RoundState).ProposalBlock.Hash()
+
+	<-voteCh // wait for prevote
```

Comment:
All new tests should use the new ensureVote() format as defined in consensus/common_test.go (and we should migrate old tests to this new format).
Just receiving events from a channel like this without validating the height/round is an absolute nightmare to debug.
We need it for all subscriptions like for propCh and newRoundCh as well.